### ISSUE TITLE
Remove Parameter / Input differentiation

### ISSUE THREAD

**Issue by bernardkwok** [autodesk-forks/MaterialX#632]
This is part of the MaterialX 1.38 Specification Update.

### Upgrade Logic

The update occurs on reading or import of a MaterialX document.
1. The upgrade will apply to any 1.37 file or if the **XMLReadOptions.applyFutureUpdates** is set to true. Note that the "updates" option is only a temporary one until 1.38 completion. **Afterwards this option will be removed**).
2. All \<parameters\> will be become \<inputs\>. If the parameter is a filename or string then it is marked with an new attribute "uniform" set to "true". Other attributes are all current set to be "true". (**completed**)
3. Special cases have not been handled yet as part of the upgrade (WIP).
4. All ShaderGen code removes any support for <parameters> (**completed**).
5. There is a way to "upgrade" / "downgrade" to/from a uniform \<input\> to a \<parameter\> explicit using the these Document methods:
```
    /// Convert Parameter Elements to Input Elements
    bool convertParametersToInputs();

    /// Convert Input Elements which are uniform to Parameter Elements
    bool convertUniformInputsToParameters();
```
**Note that these methods will be removed upon 1.38 completion.**
5. Cleanup: The \<parameter\> class will be removed. This will probably occur at the time of 1.38 release. 

**Note**: When 1.38 is complete then the upgraded file will be marked with version 1.38 until then it is marked as 1.37.

### Consumption

1. To handle this upgrade any code which gets, sets will have no affect as all parameters are now inputs. 
2. The equivalent input methods should be used instead. add and removal parameter methods should no longer be used. 
3. Any checks for RTTI to check if the type of an Element is a parameter should be replaced with **getIsUniform**(). 
4. The corresponding setter is **setIsUniform**().

Note all changes will be checked and merged into the ILM repo as progress is made (https://github.com/materialx/MaterialX).
Specifically into the 1.38 branch until release time (https://github.com/materialx/MaterialX/tree/v1.38).

Merge list:
  - #977

Don't merge list:
  - Temporary "downgrade" path: #989, #993 (only the change to make the default uniform value be true on upgrade), #996.

**Comment by niklasharrysson** [comment]
I think maybe we should change the interpretation of parameters from "a parameter port is non-connectable" into "a parameter port is uniform". _Uniform_ meaning having a constant value during frame rendering, but changeable between frames.

I agree 100% that connecting a string value upstream should be allowed. Or connecting any constant value node upstream. And also connecting from inside a nodegraph to the nodegraph's input interface, as long as the corresponding input on the graph is uniform (a parameter).

**Comment by bernardkwok** [comment]
#977 has been merged into ILM 1.38 (https://github.com/materialx/MaterialX/commit/88e2170aa3e62ae53593404d7daa878a2b76ad30)
